## Assembly pipeline
The input of the pipeline is raw Illumina paired-end data in the form of two fastq files, containing the forward and the reversed reads. On the basis of the generated genome assemblies, low quality and contaminated samples can be excluded for downstream analysis.

First, FastQC (Andrews, 2010) is used to assess the quality of the raw Illumina reads. Second, Trimmomatic (Bolger, Lohse, & Usadel, 2014) is used to remove poor quality data and adapter sequences. The sliding window option of Trimmomatic starts scanning at the 5’ end and clips the read once the average quality within the window falls below a threshold. The sliding window is the number of nucleotides over which Trimmomatic calculates an average phred quality score, a measure of the quality of the identification of the nucleobases generated by automated DNA sequencing. The Trimmomatic minlen config parameter is set to 50, this parameter is used to drop the read if the read is below a specific length.

Third, FastQC is used once more to assess the quality of the trimmed reads. Fourth, Picard  (https://broadinstitute.github.io/picard/) determines the library fragment lengths. Subsequently, the reads are assembled into scaffolds by SPAdes (Bankevich et al., 2012) by means of de novo assembly of the genome. SPAdes uses k-mers for building an initial de Bruijn graph and on following stages it performs graph-theoretical operations to assemble the genome. Kmer sizes of 21, 33, 55, 77 and 99 were used. For de novo assembly, SPAdes isolate mode is used. 

Last, quality assessment of the resulting assemblies is performed. QUAST (Gurevich, Saveliev, Vyahhi, & Tesler, 2013) is used to assess the quality of the filtered scaffolds. To assess the quality of the microbial genomes, CheckM (Parks, Imelfort, Skennerton, Hugenholtz, & Tyson, 2015) is used. CheckM calculates scores for completeness, contamination and strain heterogeneity. Bbtools (Bushnell, 2014) is used to generate scaffold alignment metrics. As a last step of the pipeline, MultiQC (Ewels, Magnusson, Lundin, & Käller, 2016) is used to summarize analysis results and quality assessments in a single report for dynamic visualization.



All input file can be put in the raw_data folder, or an input folder can be otherwise specified.

```

Dry run:
```
bash start_here.sh -n
```

Start pipeline:
```
bash start_here.sh
```
